### PR TITLE
Feat: As a user, I want to have an example merkle proof module

### LINKED ISSUE #246: [US] As a user, I want to have an example merkle proof module, So that I can see how to verify a merkle proof in my portal
As a user,
I want to have an example merkle proof module,
So that I can see how to verify a merkle proof in my portal

----

### Description

The deliverable for this user story is a contract in the `/src/example` folder of the repository that cane used as a module to verify an Ethereum ECDSA signature.

The module should implement the `AbstractModule` abstract contract and can be based on the `/src/example/CorrectModule.sol` contract.

The module accepts an `attestationPayload` and a `validationPayload`.  The `validationPayload` should be a serialised array of `bytes32` hashes which can comprises a merkle proof, and using the `attestationPayload` as a leaf node in the merkle tree and applying the proof should result in a merkle root which can be applied to some root that is also passed in as part of the `validationPayload` for comparison to the derived needle proof.

### PR DESCRIPTION
**Pull request by Builddddder**
### Related ticket

Fixes #246 

### Type of change

- [ ] Chore
- [ ] Bug fix
- [x] New feature
- [ ] Documentation update

## Check list

- [x] Unit tests for any smart contract change
- [x] Contracts and functions are documented

### CODE REVIEW

**Review by orbmis — APPROVED** [review]

File: contracts/src/examples/modules/MerkleProofModule.sol [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  function run(
+    AttestationPayload memory attestationPayload,
+    bytes memory validationPayload,
+    address /*txSender*/,
+    uint256 /*value*/
+  ) public view override {
+    bytes32[] memory proof = abi.decode(validationPayload, (bytes32[]));
+    bytes32 hash = bytes32(attestationPayload.attestationData);
+    /*
+     * @notice We send the a hardcoded third leaf to verify.
```
Comment (alainncls) [inline contracts/src/examples/modules/MerkleProofModule.sol:26]: I guess the 'a' is a typo?

File: contracts/test/example/MerkleProofModule.t.sol [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        hashes[current] = keccak256(abi.encodePacked(hashes[offset + i], hashes[offset + i + 1]));
+        current += 1;
+      }
+      offset += n;
+      n = n / 2;
+    }
+    return hashes;
+  }
+
+  function test_MerkleProofModule_MerkleProofVerifySuccess() public {
```
Comment (alainncls) [inline contracts/test/example/MerkleProofModule.t.sol:40]: The names of the two tests are swapped

**Review by alainncls — CHANGES_REQUESTED** [review]

File: contracts/src/examples/modules/MerkleProofModule.sol (hunk 1 above)
Comment (Builddddder) [reply contracts/src/examples/modules/MerkleProofModule.sol:26]: Yes, will be fixed

Comment (Builddddder) [reply contracts/src/examples/modules/MerkleProofModule.sol:26]: fixed

File: contracts/test/example/MerkleProofModule.t.sol (hunk 2 above)
Comment (Builddddder) [reply contracts/test/example/MerkleProofModule.t.sol]: fixed

File: contracts/src/examples/modules/MerkleProofModule.sol (hunk 1 above)
Comment (Builddddder) [reply contracts/src/examples/modules/MerkleProofModule.sol]: fixed

**Comment by Builddddder** [comment]
@alainncls Had changed by comments.Please review again.

**Review by alainncls — APPROVED** [review]

**Pull request merged**